Step4 - 구간 제거 기능

mins99 opened the pull request:
안녕하세요 리뷰어님!
step3의 피드백을 반영하고 step4 구현하여 리뷰 요청 드립니다.

구간 제거 기능을 추가하면서 이전 단계의 구간 조회시 정렬 문제도 수정 완료했습니다!
stream을 한 번만 사용해서 역 리스트 정렬을 해보고 싶었는데 잘 안되어서 아쉽습니다 😢

벌써 ATDD 미션의 마지막 단계네요! 익숙해질만 하면 미션이 끝나는 것 같아요ㅎㅎ
이번 단계도 잘 부탁드립니다!! 🙏🏻

mins99 committed:
test(Domain): 도메인 테스트 설명 수정

- 메서드의 동작을 적기 보다는 요구사항에 가깝게 작성하기

mins99 committed:
fix(Sections): 구간 정렬 로직 수정

mins99 committed:
docs(README): 4단계 - 구간 제거 기능 요구 사항 정리

mins99 committed:
refactor(LineService): 타입 변경 및 중복 로직 메서드 추출

- id type Long 으로 변경
- 노선 검색 메서드 추출

mins99 committed:
refactor(Section): 메서드 및 매개변수 이름 변경

mins99 committed:
refactor(LineController): id 파라미터 타입 long으로 통일

mins99 committed:
feat(removeLineStation): 지하철 구간 제거 기능 구현

- 삭제하려는 역은 시작역 / 중간역 / 종착역 중 하나의 타입에 속하므로 각 타입별 로직 구현
- 삭제하려는 역에 대한 예외 처리 추가

mins99 committed:
fix(Sections): 지하철 노선 조회 기능 수정

- 실제 구간의 순서대로 가져오지 못한 오류 수정

mins99 committed:
chore(Line): 파라미터명 수정

mins99 committed:
test(removeLineStation): 지하철 구간 제거 기능 테스트 추가

- 변수명도 일부 변경

enemfk777 commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -56,6 +93,55 @@ private Section findTargetSection(Section newSection) {
+    private boolean isContainStation(Station deleteStation) {
+        return sections.stream()
+                .flatMap(section -> section.getLineStations().stream())
+                .anyMatch(station -> station.equals(deleteStation));
```
section이 직접 station을 포함하는지 반환할 수 있다면, 내부의 모든 역들을 get으로 노출시키지 않아도 될 것 같아요 :)

enemfk777 commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -56,6 +93,55 @@ private Section findTargetSection(Section newSection) {
+    private Section findAfterDeleteTargetSection(Station station) {
+        return sections.stream()
+                .filter(currentSection -> currentSection.getUpStation().equals(station))
+                .findFirst().orElseThrow(() -> new IllegalArgumentException(NO_DATA_EXCEPTION_MESSAGE));
```
위 코멘트와 마찬가지로 내부의 station을 노출시키지 않을 수 있을 것 같아요 :)

enemfk777 commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -56,6 +93,55 @@ private Section findTargetSection(Section newSection) {
+        Section beforeSection = findBeforeDeleteTargetSection(deleteStation);
+        Section afterSection = findAfterDeleteTargetSection(deleteStation);
+
+        beforeSection.combineStationInfo(afterSection);
```
before와 after를 combine 한 새로운 객체를 add하고
이전 두 section을 지워볼 수 도 있을 것 같네요 :)
이건 민성님이 하신 방법과 비교하며 고민만 해주셔도 좋을 것 같아요

mins99 replied:
로직으로 봤을때는 a - b - c 구간에서 a - b 구간을 a - c 구간으로 변경하고 b - c 구간을 삭제하는거 보단
a - b 구간과 b - c 구간을 재조합한 a - c 구간을 추가하고 a - b 구간과 b - c 구간을 삭제하는 구조가 맞는것 같습니다!

처음에도 사실 add 하는 방식으로 구현했다가 insert가 안되는 문제가 있어서 방법을 바꾼건데
언급해주신 덕분에 cascade 옵션이 원인인걸 발견했습니다! ㅎㅎ
cascade 관련해서도 사용 타이밍(?)에 대해 정리할 수 있었습니다 👍 [(참고 링크)](https://www.inflearn.com/questions/31969)

enemfk777 commented on src/test/java/nextstep/subway/domain/SectionsTest.java:
```diff
@@ -8,72 +8,143 @@
+    @DisplayName("기존 구간에서 중간역을 삭제할 수 있다")
+    void deleteMiddleStation() {
+        // given
+        final Sections 이호선_구간 = new Sections();
+        이호선_구간.addSection(강남역_삼성역);
+        이호선_구간.addSection(삼성역_잠실역);
+
+        // when
+        이호선_구간.deleteSection(삼성역);
+
+        // then
+        assertThat(이호선_구간.getSections()).hasSize(1);
+    }
+
+    @Test
+    @DisplayName("기존 구간에 존재하지 않는 역은 삭제할 수 없다")
+    void nonExistStationDeleteError() {
+        // given
+        final Sections 이호선_구간 = new Sections();
+        이호선_구간.addSection(강남역_삼성역);
+        이호선_구간.addSection(삼성역_잠실역);
+
+        // when & then
+        assertThatIllegalArgumentException().isThrownBy(
+                () -> 이호선_구간.deleteSection(시청역)
+        );
+    }
+
+    @Test
+    @DisplayName("현재 남아있는 구간이 하나인 경우 삭제할 수 없다")
+    void lastSectionDeleteError() {
+        // given
+        final Sections 이호선_구간 = new Sections();
+        이호선_구간.addSection(강남역_삼성역);
+
+        // when & then
+        assertThatIllegalArgumentException().isThrownBy(
+                () -> 이호선_구간.deleteSection(강남역)
+        );
+    }
```
꼼꼼한 도메인 객체 테스트 💯💯

enemfk777 commented:
> 추가로, 단위 테스트도 추가하게 되면서 테스트 패키지에 대한 고민이 생겼는데요..ㅠㅠ
> 리뷰어님은 ATDD와 TDD 등등 다양한 케이스(?)의 테스트를 추가하시면 어떤 기준으로 패키지를 구분하시는지 궁금합니다!

제가 지난 PR에서 이 질문을 놓쳤네요 😭 죄송합니다.
일부러 무시한건 아니니 마음상하지 않으셨으면 좋겠습니다.🙇‍♂️

저같은 경우는 일단 유닛 테스트는 프로덕션과 동일한 레벨의 패키지로 작성하구요 (접근제한자에 따라 패키지가 달라지면 테스트가 어려워질 수도 있기 때문이에요 :))

Acceptance Test나 통합 테스트 같은 경우는 따로 패키지를 분리하고, 그 아래서 도메인별로 나눠서 하위 패키지를 만든 후 진행하는 편입니다!

enemfk777 reviewed (changes requested):
안녕하세요 민성님 🙇‍♂️
리뷰가 늦어서 죄송합니다 😭
4단계를 손쉽게 진행해주셨네요 💯
지난 단계의 마지막 코멘트도 반영해주셔서 감사합니다!!

몇가지 소소한 피드백이 있으니 반영 후 다시 한 번 요청 부탁드립니다!!

mins99 committed:
refactor: 지하철 구간 제거 기능 개선

- A-B-C 구간의 B역 삭제시 A-C 구간을 추가 후 기존 구간을 삭제하는 로직으로 변경
- getter를 최대한 삭제
1) distance 거리 계산 로직
2) sections의 upstation, downstation 비교 로직

mins99 commented:
안녕하세요 리뷰어님!
평일에 시간내시기 어려우실텐데 늦은 시간까지 리뷰하시느라 고생이 많으십니다 ㅠㅠ 
해주시는 것만으로도 저는 매우 감사하니 죄송해 하지 않으셔도 됩니다!!

남겨주신 피드백 보고 코드를 다시 보니 정말 개선 가능한 부분들이 많더라구요! ㅎㅎ 시간이 많이 걸리기는 했지만 즐거웠습니다 🥰
추가로 step3에서 피드백주셨던 Distance 관련해서도 수정해보았습니다! https://github.com/next-step/atdd-subway-admin/pull/687#discussion_r884987802

수정한 부분이 많아서 오히려 또 수정을 해야할지도 모른다는 걱정이 조금있지만..!
확인 부탁드릴게요! 🙇🏻

mins99 commented on src/main/java/nextstep/subway/domain/Section.java:
```diff
@@ -66,38 +73,39 @@ public void setLine(final Line line) {
-            changeDistance(section.getDistance());
+        if (upStation.equals(section.upStation)) {
+            this.upStation = section.downStation;
+            this.distance.subtractDistance(section.distance);
         }
-        if (downStation.equals(section.getDownStation())) {
-            changeDownStation(section.getUpStation());
-            changeDistance(section.getDistance());
+        if (downStation.equals(section.downStation)) {
+            this.downStation = section.upStation;
+            this.distance.subtractDistance(section.distance);
         }
     }
```
getter를 없애고자 하다보니 다음의 방식을 사용했는데, 사실 처음! 써본거라 괜찮은 방법인지 잘 모르겠네요..;;
해당 방식으로도 많이 쓰나요..?(클래스의 메서드 파라미터를 동일 클래스로 쓰기)
private 접근 제어자가 동일 클래스 내에서만 접근이 가능해서 해당 코드가 가능한것 같긴 한데
인스턴스가 달라도 접근이 가능한가요?
관련 내용을 찾지를 못해서 혹시 아시는 문서가 있다면 공유 해주시면 감사하겠습니다!

enemfk777 replied:
넵 해당 방식으로 많이 씁니다 :)
동일 클래스 내이기때문에 사실 public 메서드인 getter를 거칠 필요가 없죠 :)

enemfk777 reviewed (approved):
안녕하세요 민성님 :)
피드백 잘 반영해주셨네요!!
지난 단계의 피드백까지 잊지않고 반영해주셔서 감사합니다 🙇‍♂️

인수테스트 주도 개발 미션은 여기까지 입니다 :)
저와 주고 받았던 피드백이 앞으로 미션을 진행하시는데 도움이 된다면 좋겠네요 :)
그럼 마지막까지 완주하시길 응원하겠습니다! 💪
